ast/desugar/Desugar.cc

```diff
@@ -2084,7 +2084,10 @@ ExpressionPtr node2TreeImpl(DesugarContext dctx, unique_ptr<parser::Node> what)
                 result = std::move(res);
             },
             [&](parser::Backref *backref) {
-                auto res = unsupportedNode(dctx, backref);
+                // Desugar backref expression as T.let(nil, T.nilable(String))
+                auto value = MK::Nil(loc);
+                auto type = MK::Nilable(loc, MK::Constant(loc, core::Symbols::String()));
+                ExpressionPtr res = MK::Let(loc, std::move(value), std::move(type));
```

Comment:
Using `T.let` is going to cause the variable to be pinned even when the user did not explicitly pin the variable.

Can we instead desugar this to a custom method like `<Magic>.<regex-backref>(...)`, and make that method's signature say `T.nilable(String)`?

Here's a test case for you

```ruby
# typed: true
extend T::Sig

sig {returns(T.nilable(String))}
def magic_backref; end

/foo/.match("foo") do |m|
  x = $&
  x = false

  y = magic_backref
  y = false
end
```

These two things should behave the same, but this change treats them differently. (This PR causes an error on `x = false` but no error on `y = false`).

This will also be better long term. At runtime, the value is not actually `nil`. Modeling the runtime behavior is important for us too.

Reply:
Look in GlobalState.cc for how to define new <Magic> methods.

Reply:
Thank you! Will dig back into this.